### DISCUSSION TITLE
Limiting agent interactions to once per step - updating both agents

### DISCUSSION THREAD

**Discussion by zeyus** [FLAMEGPU/FLAMEGPU2#888] [Q&A] [answered]
I've changed up what I want to do with the ABM, and basically we're going to do some variants on the prisoner's dilemma (with environmental and reproduction costs).

What I'm wondering is how conceptually to best limit agent pair interactions to once per step. I'm using `flamegpu::MessageArray2D` and the moore wrap (currently radius 1). 

Each agent will output a message with its strategy, both agents should be affected by the interaction, but two agents in adjacent positions should play each other once, yet be updated with the reward or penalty for the interaction.

I'm trying to avoid keeping a second copy of the array in a macro property and host function to track interactions, but maybe this is the only way? 

I guess the problem I'm facing is fully wrapping my head around the scope of all the variables and where they're allowed to be updated, for example if these were objects, I'd just keep a list of interactions per agent for the current step, and, once they played together they would not attempt again.

Here's the basic outline of what I'm doing:

```cuda
FLAMEGPU_AGENT_FUNCTION(interact, flamegpu::MessageArray2D, flamegpu::MessageNone) {
    const unsigned int my_x = FLAMEGPU->getVariable<unsigned int>("x_a");
    const unsigned int my_y = FLAMEGPU->getVariable<unsigned int>("y_a");
    const unsigned int interaction_radius = FLAMEGPU->environment.getProperty<unsigned int>("max_play_distance");
    
    const float reproduction_threshold = FLAMEGPU->environment.getProperty<float>("reproduce_min_energy");
    const float reproduction_cost = FLAMEGPU->environment.getProperty<float>("reproduce_cost");

    float my_energy = FLAMEGPU->getVariable<float>("energy");
    
    // iterate over all cells in the neighbourhood
    // this also wraps across env boundaries.
    for (auto &message : FLAMEGPU->message_in.wrap(my_x, my_y, interaction_radius)) {
        flamegpu::id_t local_competitor = message.getVariable<flamegpu::id_t>("id");
        // play with the competitor

    }
    if (my_energy <= 0.0) {
        return flamegpu::DEAD;
    }
    if (my_energy > reproduction_threshold) {
        // spawn child in a free adjacent cell
        FLAMEGPU->setVariable<float>("energy", my_energy - reproduction_cost);
        return flamegpu::ALIVE;
    }
    return flamegpu::ALIVE;
}
```

**Comment by zeyus** [comment, marked as answer]
Continuing to think about this, it may help (but I still can't imagine the way to stop race conditions) to increase the number of layers, and make the appropriate layers conditional, as well as adding a couple of extra agent states, (e.g. ready, played_neighbours...)

And the layer structure something like: 

1. Send out search messages
2. Play with nearby players and die if energy is too low
3. conditionally move if no play occurred (i.e. no neighbours) 
4. conditionally reproduce if energy is high enough

**Reply by zeyus** [reply]
for example I just tried this but it errors with read/atomic write cannot be mixed in the same layer

```cuda
FLAMEGPU_AGENT_FUNCTION(interact, flamegpu::MessageArray2D, flamegpu::MessageNone) {
    const unsigned int my_x = FLAMEGPU->getVariable<unsigned int>("x_a");
    const unsigned int my_y = FLAMEGPU->getVariable<unsigned int>("y_a");
    const unsigned int my_grid_index = FLAMEGPU->getVariable<unsigned int>("grid_index");

    // const unsigned int max_agents = FLAMEGPU->environment.getProperty<unsigned int>("max_agents");
    // const unsigned int env_max = FLAMEGPU->environment.getProperty<unsigned int>("env_max");
    
    const unsigned int interaction_radius = FLAMEGPU->environment.getProperty<unsigned int>("max_play_distance");
    
    const float reproduction_threshold = FLAMEGPU->environment.getProperty<float>("reproduce_min_energy");
    const float reproduction_cost = FLAMEGPU->environment.getProperty<float>("reproduce_cost");

    float my_energy = FLAMEGPU->getVariable<float>("energy");
    // replace dimensions with python string formatting so agent count can vary
    auto playspace = FLAMEGPU->environment.getMacroProperty<unsigned int, 16384, 16384>("playspace");
    // iterate over all cells in the neighbourhood
    // this also wraps across env boundaries.
    for (auto &message : FLAMEGPU->message_in.wrap(my_x, my_y, interaction_radius)) {
        flamegpu::id_t local_competitor = message.getVariable<flamegpu::id_t>("id");
        unsigned int opponent_grid_index = message.getVariable<unsigned int>("grid_index");
        // play with the competitor
        if (playspace[opponent_grid_index][my_grid_index] == 0) {
            // we haven't played before
            playspace[my_grid_index][opponent_grid_index]++;
        } else {
            // we have played before, and this is the second of two possible interactions
            // do nothing except reset grid
            playspace[opponent_grid_index][my_grid_index]--;
        }

    }
    if (my_energy <= 0.0) {
        return flamegpu::DEAD;
    }
    if (my_energy > reproduction_threshold) {
        // spawn child in a free adjacent cell
        FLAMEGPU->setVariable<float>("energy", my_energy - reproduction_cost);
        return flamegpu::ALIVE;
    }
    return flamegpu::ALIVE;
}
```

**Reply by Robadob** [reply]
> for example I just tried this but it errors with read/atomic write cannot be mixed in the same layer

Yes the atomic methods for macro properties are setup to prevent most race conditions.

You should be able to restructure what you're doing to be permitted. E.g.

```c
        if (++playspace[my_grid_index][opponent_grid_index] != 1) {  // pre-increment is marginally cheaper than post-increment
            playspace[my_grid_index][opponent_grid_index] -= 2;  // Small chance this needs to be two separate decrements
        }
```

*Note: Dependent on context, this may still contain a race condition, have a think about it.*

I'm still trying to get a clear understanding of what you're trying to do, but it feels like a problem which might work better with a submodel made up of several layers. Iterative algorithms are best for performing parallel conflict resolution (The SugarScape example demonstrates one of these) It's very easy to create unwanted bias when dealing with macro properties.

Another possible 'hack' is creating a rule whereby the agent with the greater ID always performs the interaction or similar. This kind of rule can be used to implicitly block both sides performing the check.

I'll try to post a clearer suggestion once I've had a chance to get my head around what you're trying to do.

**Reply by Robadob** [reply]
Can the interactions be one sided? The sample code you're provided doesn't detail the interaction. But logically if an agent requires energy to interact, if they run out of energy they can't interact with any more neighbours?

My rough thought would be:
* Submodel (this can perform as many iterations required to resolve all interactions, then exit returning to main model)
* Every agent generates a random float for this round, which they store locally and attach to a message (with their ID and energy)
* Each agent iterates messages, and does some basic psuedo-rng with the two input floats (see end of this comment for an example) to generate a score for each valid interaction (has energy, has not occurred before, is in range, etc). Interaction with highest score is then put forwards by each agent (2nd message output?).
* Next agent function then does something to decide which interactions take place (if both sides don't agree)
* Agent with lowest ID, updates a macro property to mark the interaction as complete/nolonger valid.
* Agents that have no interactions left set some internal variable, You can then check whether this is set (or zero) for all agents in an exit function.


But I'd really need to know more about the interaction to design it better, ideally you structure it so as many interactions as possible can be performed per iteration, but I would need to better understand how 'energy' is consumed for that.

It can be useful to mock out such algorithms in serial before you try and model them, so you can ensure there's no unintended bias. You can find such mocking I did [here](https://gist.github.com/Robadob/ddd974e47bbb50aca7c9a5342b42588b) when needing to develop a parallel suitable cell bonding algorithm. We only ever implemented that model in our modified version of FLAMEGPU 1 though, it was removed as redundant before we ported the model to FLAMEGPU 2. (To summarise this bonding algorithm, Every cell needs to create a maximum of 6 mutual bonds with nearby cells which satisfy some additional conditions. After these have been created, the bonds would apply a force to each bonded cell pair).

**Reply by zeyus** [reply]
> Another possible 'hack' is creating a rule whereby the agent with the greater ID always performs the interaction or similar. This kind of rule can be used to implicitly block both sides performing the check.

That hack might actually solve the issue, as long as the game results are saved and the next layer updates both agents (which is definitely possible).

> Can the interactions be one sided? But logically if an agent requires energy to interact, if they run out of energy they can't interact with any more neighbours?

Yeah this is a fair point, the condition will still exist where an unfair matching of an agent with a dead agent (with the alive agent potentially losing a match that should never have occurred)

> My rough thought would be: ...

This is really useful thanks, I think this approach could work well.

> But I'd really need to know more about the interaction to design it better, ideally you structure it so as many interactions as possible can be performed per iteration, but I would need to better understand how 'energy' is consumed for that.

What I want to achieve as the starting point is actually fairly simple conceptually, it's just my lack of experience with this way of working and my obsession with making the most optimal code  that I'm capable of haha. 

Energy can be gained by one agent and lost from another, or both zero, or both gain depending on their strategies, as well as a adjustable "noise" level (which would negate their choice of cooperation or defecting). Energy would also be spent on reproducing, and also if there are too many agents, there would also potentially be a cost (but this isn't required for the first implementation, I'm really trying to get a flexible skeleton MVP up and running to extend from there).
The basic model will start as an implementation of this NetLogo model: https://www.comses.net/codebases/5500/releases/1.0.0/

```logo
;;try to copy a simple spatial PD game with mobile agents

;;turtle variables
turtles-own[
 energy ;turtles accumulated energy
 played? ;has the turtle played this round?
]

;;global parameters
globals[
  reproduce-threshold
  reproduce-cost
  payoff-CC
  payoff-DC
  payoff-DD
  max-energy
]


;;SETUP PROCEDURES

to setup
  clear-all
  ask patches [set pcolor white]
  initialize-variables
  create-agents num-agents init-coop-freq
  reset-ticks
end

to initialize-variables
  ;;initialize all variables
  set reproduce-threshold 100
  set reproduce-cost 50
  set payoff-CC 3
;  set payoff-CD -1
  set payoff-DC 5
  set payoff-DD 0
  set max-energy 150
end

to create-agents [n coop-freq]
  create-turtles n[
    ifelse random-float 1 < coop-freq
    [set color blue]
    [set color red]
    set shape "circle"
    set size 1.0
    move-to one-of patches with [not any? other turtles-here]
    set energy (random 50 + 1)
  ]
end




;;BEHAVIOR PROCEDURES

to go
  if not any? turtles [ stop ] ;stop the simulation if everyone is dead
  ask turtles [set played? false]
  ask turtles[
    if energy <= 0 [die] ;;die if not enough energy

    ;;play PD game and collect payoffs
    if (not played?)[ interact ] ;;find coplayer

    ;;reproduce
    if (energy >= reproduce-threshold and (count turtles < carrying-capacity))
      [reproduce]

    ;;energy loss
    set energy (energy - cost-of-living)
    if energy > max-energy [ set energy max-energy ]

    ;;movement
    if (not played?) [ move2 ];;if didn't find coplayer, move
  ]
  tick
end



;;Find an interaction partner and play PD game with them
to interact
  let partner (one-of ((turtles-on neighbors) with [played? = false]))
  ;;the rest only runs if a partner is found
  if partner != nobody [
    set played? true
    ask partner [set played? true]

    ifelse color = blue ;if I'm a cooperator
    [
      ifelse ([color] of partner) = blue
      [
        set energy (energy + payoff-CC)
        ask partner [set energy (energy + payoff-CC)]
      ]
      [
        set energy (energy + payoff-CD)
        ask partner [set energy (energy + payoff-DC)]
      ]
    ]
    [ ;if I'm a defector
      ifelse ([color] of partner) = blue
      [
        set energy (energy + payoff-DC)
        ask partner [set energy (energy + payoff-CD)]
      ]
      [
        set energy (energy + payoff-DD)
        ask partner [set energy (energy + payoff-DD)]
      ]
    ]
  ]
end


;;move to a random new location if no one to play with
;;for now, just move to any available space in the Moore neigborhood.
to move
  let new-patch (one-of neighbors with [ not any? turtles-here ])
  if (new-patch != nobody)
  [
    move-to new-patch
  ]
end

;;movement more in line with original Am Nat model
;;agent chooses square to move to, and only then checks if it's occupied
to move2
  let new-patch (one-of neighbors)
  if (not any? turtles-on new-patch)
  [
    move-to new-patch
  ]
end


;;attempt to place a duplicate agent in a neighboring cell
to reproduce
  let new-patch (one-of neighbors with [ not any? turtles-here ])
  if (new-patch != nobody)
  [
    set energy (energy - reproduce-cost)
    hatch 1 [
      set energy reproduce-cost
      move-to new-patch
      set played? false
  ]
  ]
end

;;random agents lose energy if population exceeds carrying capacity
to cull-herd
  if (count turtles > carrying-capacity)
  [
    ask one-of turtles [
      set energy ( energy - 10 )
    ]
  ]
end


;;for plotting

to-report rep-coop-freq
  let Nc (count turtles with [ color = blue ])
  let N (count turtles)
  report Nc / N
end



;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Copyright 2017 Paul Smaldino
;; See info tab for details
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
```

**Reply by zeyus** [reply]
and my work in progress: 
https://github.com/zeyus/SoCultABM/blob/OL_PD/src/pd/model.py

Also, I really appreciate the advice, I'm definitely not looking for you to write the code for me, but was just struggling with trying to avoid unnecessary, unwanted, or detrimental interactions (without having to make the whole thing run in host code, otherwise I'm not taking advantage of FLAMEGPU at all!)

**Reply by Robadob** [reply]
> struggling with trying to avoid unnecessary, unwanted, or detrimental interactions (without having to make the whole thing run in host code, otherwise I'm not taking advantage of FLAMEGPU at all!)

Converting serial "conflict resolution" or "symmetry breaking" behaviours to parallel isn't particularly straight forwards, and is not something we expect most modellers to have a strong grasp of. We've yet to come up with a clear plan of how to walk users through taking a serial algorithm (where generally you just build a list of all possible interactions, shuffle it, and operate on them in order) and converting it to a parallel algorithm.

I'll try and have a look through the code you posted later this evening/tomorrow. I should also get to your PR on Monday (I've been on annual leave past couple of weeks, so haven't had the code open to do a quick build).

**Reply by Robadob** [reply]
Ah so looking at that code

```
;;Find an interaction partner and play PD game with them
to interact
  let partner (one-of ((turtles-on neighbors) with [played? = false]))
  ;;the rest only runs if a partner is found
  if partner != nobody [
    set played? true
    ask partner [set played? true]
```

This suggests You essentially want to find a single mutually agreed partner each round. That makes it much simpler, essentially a simplified version of the bonding algorithm (1 bond rather than 6) I linked python code to in my previous comment.

So the above structure should work fine. Instead of coming up with a complicated way to decide which bond to use when agents don't agree. It should be much easier to simply have agents only bond when they do agree, just keep an eye on how many iterations the submodel is taking to max out the bonds.

I may have time to put together an example of a simple bonding submodel this week. I expect it's something my boss would like us to have as an example, if it more clearly fits general requirements than the existing sugarscape model. (One long-term idea has been to provide some general common submodels for things such as cell-cell interaction to assist users, if we can find a clean way of making them usable)

I haven't looked further into the code you posted, as I don't think I need to understand more of it right now.

**Reply by zeyus** [reply]
Hey thanks for that, I think we want to model a slightly more flexible version of it, and I've made a sketch of the flow here. I'm also 100% happy for whatever the end model we use to be copied or altered as required for an example, I'm trying to make it flexible, both for the purposes of avoiding rework and allowing for multiple runs with changing parameters (probably going to try and use the runplan framework) as well as providing a repo that anyone can use to run a 2D spatial population sim from.
![serial_to_parallel_pd-General flow drawio](https://user-images.githubusercontent.com/75656/182042246-7db3c113-348f-44c8-b5c0-49c9698afcad.png)

> and converting it to a parallel algorithm.

This is what I've been struggling with, I've worked with parallelized applications before and implemented it for massive performance gains, but I'm used to working with signaling or thread locking, where in this case I would lock an agent while it interacts with those around it, and update it's values, including which other agents it has interacted with, but any agents outside of that radius of interaction would continue to interact with unlocked agents.

> I may have time to put together an example of a simple bonding submodel this week.

That would be interesting to see, I'll be working on this ABM over the next few days and really need to get it done in order to have some results and data from simulations. I'm hoping that I can piece together something great from the advice you've provided here, I'm not fully sure on how implementing the submodels will work to avoid unnecessary interactions, but I will take a deeper look at the sugarscape model tomorrow and see what I can come up with.

**Reply by zeyus** [reply]
How would the sugarscape model scale (using an agent for each grid space, even if it's unoccupied) compared to representing each cell as an individual agent?

**Reply by Robadob** [reply]
That's a question of how GPUs operate.

A quick summary of GPU theory is that they follow the Single Instruction Multiple Thread (SIMT) paradigm. When you launch 1000 threads, they're grouped into 'warps' of 32 neighbouring threads. Each instruction is issued for a a single warp of 32 threads simultaneously. *(GPUs have multiple streaming multiprocessors, which can each operate on 10s of warps concurrently to hide latencies which is how modern GPUs can support 50,000+ threads in flight concurrently, but that's kind of irrelevant to this point)*

A thread in that warp either executes the instruction or goes inactive and ignores it (e.g. if there's divergence, such as an if-statement branching).

Optimal GPU code therefore, has every thread within a warp following the exact same code without divergence. Whereas, the worst case would be all 32 threads following completely different code paths (things will take ~32x longer!).

I would therefore expect that having a thread per agent, rather than a thread per grid cell is optimal in terms of performance as it would minimise divergence and achieve better device utilisation. If the agents are sorted to maximise memory locality/reduce divergence you might get a slight improvement too (this makes a big difference with our spatial messaging).

In the case of SugarScape however, the grid cells all(?) contain sugar which much grow back each step, and the agents are treated as 'virtual agents' living in the grid cells and able to move about. Ideally you would model the sugar grid and the 'agents' as two separate agents, but then alot more communication would be required (we could alternatively store the sugar in an environment macro property with flamegpu2). Given the simplicity of the model, we made the decision to keep the agents 'virtual'.

The example models are small 'toy' models, they've not been heavily calibrated or optimised.

Furthermore, we don't yet have a proper way to visualise cellular automata (the current visualisations are a bunch of stacked cubes). So it's necessary to have an agent per cell if you want all cells to be visualised (a 1 sided plane square + clear colour code work though tbh).

**Reply by Robadob** [reply]
Have just invited you to [private repo with a 2D discrete example of the Bonding model](https://github.com/FLAMEGPU/bonding-submodel-example) I described above. It's a bit different to your use-case as every cell is currently occupied (and in the C++ API), but should give you something to work from.

~~Regardless of the agent count it seems to resolve all bonds in 9-11 iterations of the submodel.~~ 
**Edit:** I fixed some bugs, and applied a more random (but agreed upon) `bond_score` algorithm. It now resolves in 7-8 steps (I haven't bothered updating the speeds below, so should be ~20% faster than stated).

Will probably make it public (or move it to the main repo's examples) once I've had a chance to discuss it with the wider team.

On my old laptop's 980m, and a (Release/Seatbelts=OFF/Visualisation=ON build).
~~65,000 agents is running at 2.0 steps/second (bit slower than I would expect), but so is 1 million agents 🤷‍♂️. So much faster than I expect NetLogo would be at that scale,~~
**Edit::** Just realised, the vis code I copypasted locks speed to 2 steps/second. 1 million agents actually runs at 9.4 steps/second, and 65k agents ~56 steps/second.

(a step is all agents bonding and then playing rock-paper-scissors with their partner)

A couple of screenshots to crudely show that bonds change each step (red means the cell failed to bond).
![image](https://user-images.githubusercontent.com/742154/182474329-20bce406-114f-4c28-9532-011ead2b9d49.png)
![image](https://user-images.githubusercontent.com/742154/182474352-77dfc826-54ed-4cb2-9da7-7e81ea61e83e.png)

**Reply by zeyus** [reply]
Hey @Robadob thanks, this is amazing, I think I'll implement a variant of that bonding algorithm in my code, and it should actually address the issue I am facing in https://github.com/FLAMEGPU/FLAMEGPU2/discussions/894 , as the requests will come from the space where the agent is, rather than trying to output a message at the opponent's space (even though it should be possible in theory)...Also I think the random number is a good idea, rather than relying on the agent's ID. Right now I'll just try and get it done, and not stress too much about optimization, although it's one of my favorite parts of coding haha. I think with the method you've outlined for bonding, it wouldn't even be necessary to have a movement submodel, making the initial neighbour broadcast also redundant, but I'll leave them for now, and hopefully the function condition will prevent too much unnecessary overhead, though it would probably be worth removing 2x message out/in steps if it's not needed at some point.

**Reply by Robadob** [reply]
I wouldn't worry too much about overheads (especially until you have it working, and know whether you need more performance).

Discrete models are incredibly cheap compared to continuous space models. This is largely a consequence of a significantly reduced volume of communication (e.g. messages read) and a more balanced workload.
